controllers/state_of_the_world.go

```diff
@@ -39,9 +42,17 @@ var (
 	operatorNamespace  = env.GetString("OPERATOR_NAMESPACE", "kuadrant-system")
 )
 
+type PolicyMachineryController struct {
+	isGatewayAPIInstalled    bool
+	isEnvoyGatewayInstalled  bool
+	isIstioInstalled         bool
+	isCertManagerInstalled   bool
+	isConsolePluginInstalled bool
+}
+
 //+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gatewayclasses,verbs=list;watch
 
-func NewPolicyMachineryController(manager ctrlruntime.Manager, client *dynamic.DynamicClient, logger logr.Logger) *controller.Controller {
+func (c *PolicyMachineryController) Controller(manager ctrlruntime.Manager, client *dynamic.DynamicClient, logger logr.Logger) *controller.Controller {
```

Comment:
I would probably keep this as a direct constructor that returns a `*controller.Controller`, instead of a method of a struct that now needs to be known at `main.go`.

Additionally, I would make `PolicyMachineryController` into `policyMachineryControllerOptionsBuilder`. Make it an internal thing used by `NewPolicyMachineryController`. Name is probably too long BTW.

The builder could implement methods such as `getGatewayAPIOptions() ([]controller.ControllerOption, error)`, `getEnvoyGatewayoptions() ([]controller.ControllerOption, error)` and `buildReconciler() controller.ReconcileFunc`.

I think this can help breaking down the constructor into smaller steps and avoid any temptation of doing `&PolicyMachineryController{}` and getting something that actually is not a Policy Machinery Controller yet.

WDYT?

Reply:
Yeah, this sounds a lot nicer, I'll push a change for this 👍